test/jdk/javax/net/ssl/TLSv13/SSLEngineNoServerHelloClientShutdown.java

```diff
@@ -0,0 +1,270 @@
+        clientEngine.setEnabledProtocols(new String[] {"TLSv1.3"});
+        return clientEngine;
+    }
+
+    /*
+     * Configure the server side engine.
+     */
+    protected SSLEngine configureServerEngine(SSLEngine serverEngine) {
+        serverEngine.setUseClientMode(false);
+        serverEngine.setNeedClientAuth(true);
+        return serverEngine;
+    }
+
+    public static void main(String[] args) throws Exception {
+        new SSLEngineNoServerHelloClientShutdown().runTestUserCancelled();
+    }
+
+    //
+    // Private methods that used to build the common part of the test.
+    //
+
+    private void runTestUserCancelled() throws Exception {
+        SSLEngineResult clientResult;
+        SSLEngineResult serverResult;
+
+        log("=================");
+
+        // client wrap
+        // produce client_hello
+        log("---Client Wrap client_hello---");
+        clientResult = clientEngine.wrap(clientOut, cTOs);
+        logEngineStatus(clientEngine, clientResult);
+        runDelegatedTasks(clientEngine);
+
+        cTOs.flip();
+
+        // server unwrap
+        // Consume client_hello
+        log("---Server Unwrap client_hello---");
+        serverResult = serverEngine.unwrap(cTOs, serverIn);
+        logEngineStatus(serverEngine, serverResult);
+        runDelegatedTasks(serverEngine);
+
+        cTOs.compact();
+
+        // server wrap
+        // produce server_hello
+        log("---Server Wrap server_hello---");
+        serverResult = serverEngine.wrap(serverOut, sTOc);
+        logEngineStatus(serverEngine, serverResult);
+        runDelegatedTasks(serverEngine);
+        sTOc.clear();  // SH packet went missing.  Timeout on Client.
+
+        // server wrap
+        // produce other outbound messages
+        log("---Server Wrap---");
+        serverResult = serverEngine.wrap(serverOut, sTOc);
+        logEngineStatus(serverEngine, serverResult);
+        runDelegatedTasks(serverEngine);
+        sTOc.clear();  // CCS packet went missing.  Timeout on Client.
+
+        // server wrap
+        // produce other outbound messages
+        log("---Server Wrap---");
+        serverResult = serverEngine.wrap(serverOut, sTOc);
+        logEngineStatus(serverEngine, serverResult);
+        runDelegatedTasks(serverEngine);
+        sTOc.clear();  // EE/etc. packet went missing.  Timeout on Client.
+
+        // client wrap
+        // *Yawn*  No response.  Shutdown client
+        log("---Client Wrap closeOutbound---");
+        clientEngine.closeOutbound();
+
+        // Sends an unencrypted user_cancelled
+        clientResult = clientEngine.wrap(clientOut, cTOs);
+        logEngineStatus(clientEngine, clientResult);
+        runDelegatedTasks(clientEngine);
+
+        cTOs.flip();
+
+        // Server unwrap should process an unencrypted 2 byte packet,
+        log("---Server Unwrap user_cancelled alert---");
+        serverResult = serverEngine.unwrap(cTOs, serverIn);
```

Comment:
does that mean that we don't throw a SSLHandshakeException on receipt of close_notify in the middle of the handshake?

Reply:
Our own client code is unable to produce `close_notify` during the handshake, it sends `user_cancelled` as per RFC. If `close_notify` alert is received during handshake and it doesn't follow the `user_cancelled` alert, then we throw UNEXPECTED_MESSAGE SSLHandshakeException.

Reply:
Please disregard my last comment. Our own client code does produce `close_notify` after `user_cancelled` during the handshake, as it SHOULD per RFC. I've updated the test to reflect that.